ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,895 @@
+    List<IndexEntry> copyCandidates =
+        getValidIndexEntries(indexSegmentStartOffset, allIndexEntries, startOffsetOfLastIndexSegmentForDelete);
+    for (IndexEntry copyCandidate : copyCandidates) {
+      IndexValue copyCandidateValue = copyCandidate.getValue();
+      // search for duplicates in srcIndex if required
+      if (duplicateSearchSpan != null) {
+        IndexValue possibleDuplicate = srcIndex.findKey(copyCandidate.getKey(), duplicateSearchSpan);
+        if (possibleDuplicate != null) {
+          if (possibleDuplicate.isFlagSet(IndexValue.Flags.Delete_Index)) {
+            // copyCandidate is surely a duplicate because srcIndex contains a DELETE index entry
+            continue;
+          } else if (!copyCandidateValue.isFlagSet(IndexValue.Flags.Delete_Index)) {
+            // copyCandidate is a duplicate because it is a PUT entry and so is the possible duplicate.
+            continue;
+          }
+        }
+      }
+      // search for duplicates in tgtIndex if required
+      if (!checkAlreadyCopied || !alreadyExistsInTgt(copyCandidate.getKey(), copyCandidateValue)) {
+        logger.trace("Adding index entry for {} because it is valid and does not already exist",
+            copyCandidate.getKey());
+        indexEntriesToCopy.add(copyCandidate);
+      }
+    }
+    // order by offset in log.
+    Collections.sort(indexEntriesToCopy, PersistentIndex.INDEX_ENTRIES_OFFSET_COMPARATOR);
+    return indexEntriesToCopy;
+  }
+
+  /**
+   * Gets all the valid index entries in the given list of index entries.
+   * @param indexSegmentStartOffset the start {@link Offset} of the {@link IndexSegment} that {@code allIndexEntries} is
+   *                                from.
+   * @param allIndexEntries the list of {@link IndexEntry} instances from which the valid entries have to be chosen.
+   * @param startOffsetOfLastIndexSegmentForDelete the start {@link Offset} of the index segment up until which delete
+   *                                                records are considered applicable.
+   * @return the list of valid entries picked from {@code allIndexEntries}.
+   * @throws StoreException if {@link BlobReadOptions} could not be obtained from the store for deleted blobs.
+   */
+  private List<IndexEntry> getValidIndexEntries(Offset indexSegmentStartOffset, List<IndexEntry> allIndexEntries,
+      Offset startOffsetOfLastIndexSegmentForDelete) throws StoreException {
+    // TODO: move this blob store stats
+    boolean deletesInEffect = startOffsetOfLastIndexSegmentForDelete != null
+        && indexSegmentStartOffset.compareTo(startOffsetOfLastIndexSegmentForDelete) <= 0;
+    List<IndexEntry> validEntries = new ArrayList<>();
+    for (IndexEntry indexEntry : allIndexEntries) {
+      IndexValue value = indexEntry.getValue();
+      if (value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+        // DELETE entry. Always valid.
+        validEntries.add(indexEntry);
```

Comment:
If the corresponding PUT record ends up getting included in `validEntries` later, I would prefer if you ensure that it goes before this DELETE record, so that the sorting that you do later is only an optimization and not required for correctness.